### PR TITLE
Greatly improve the test framework

### PR DESCRIPTION
**Pull request by Nadrieril**
Also reworked the datatypes used to describe task changes

### CODE REVIEW

File: src/stable_marriage.rs [hunk 1]
```diff
@@ -1,96 +1,234 @@
...
+use std;
+use std::cmp::Ordering;
 
-    // For efficiency: most preferred last for men, most preferred first for women
-    is_man: bool,
-    prefs: Vec<usize>,
-}
+pub trait Matcher {
+    type Item;
+    type Target;
```
Comment (Ekleog) [inline src/stable_marriage.rs]: In this file there is currently a mix between `Item` / `Target` and `Man` / `Woman`. If I understand correctly (and the need for me to say that proves the issue 😛), `Item` is basically `Man` (minus some computed metadata) and `Target` is `Woman` (minus some computed metadata).

I think it'd be better to rename these elements based on this fact, if I'm not mistaken?

File: src/stable_marriage.rs [hunk 2]
```diff
@@ -1,96 +1,234 @@
...
+        };
+
+        if let Some(woman) = men_matcher.find_perfect_match(&man.data, &mut women) {
+            woman.current_is_perfect = true;
+            woman.replace_match(man);
+            continue;
+        }
+        man.prefs = men_matcher.compute_preference_list(&man.data, &women, women_matcher);
+
+        // Loop while we still hold an engageable man
```
Comment (Ekleog) [inline src/stable_marriage.rs]: Would reword for “Loop while the man we hold is still engageable”, so as to avoid the confusion I fell in, which is thinking that we're going to loop over men here.

File: src/stable_marriage.rs [hunk 3]
```diff
@@ -1,96 +1,234 @@
...
-pub fn stable_marriage(
-    men: Vec<Vec<usize>>,
-    women: Vec<Vec<usize>>,
-) -> (Vec<Option<usize>>, Vec<Option<usize>>) {
-    let (mut men, mut women) = init_matching(men, women);
-
-    saturate_matching(&mut men, &mut women);
-
-    (extract_matching(men), extract_matching(women))
+// This algorithm favors men.
```
Comment (Ekleog) [inline src/stable_marriage.rs]: Hmm… I'm not 100% sure, but it more or less looks like this is no longer a stable matching, but a greedy matching from the man's perspective. Or are the two actually identical? (thinking esp. of a man who'd be evicted by another man later in the list, who'd never have a chance to meet the woman after his first in his list)

File: src/stable_marriage.rs [hunk 4]
```diff
@@ -1,96 +1,234 @@
...
+    ) -> Option<&'a mut Woman<Self>> {
+        targets
+            .iter_mut()
+            .filter(|x| x.current_match.is_none())
+            .find(|x| self.is_perfect_match(item, &x.data))
     }
+}
 
-    (men, women)
+struct Man<P: Matcher + ?Sized> {
```
Comment (Ekleog) [inline src/stable_marriage.rs]: I *knew* you'd give in and eventually have separate types for men and women :stuck_out_tongue_closed_eyes:

File: src/stable_marriage.rs [hunk 5]
```diff
@@ -1,96 +1,234 @@
...
+            // Cheat, because we know which targets will be used
+            self.0[*item].iter().cloned().rev().collect_vec()
+        }
+    }
+
+    fn numbers() -> impl Iterator<Item = usize> {
+        itertools::unfold(0, |x| {
+            let v = *x;
+            *x += 1;
+            Some(v)
```
Comment (Ekleog) [inline src/stable_marriage.rs]: Hmm… Isn't this `(0..)`?

File: src/task_change.rs [hunk 6]
```diff
@@ -413,97 +473,119 @@ pub fn compute_changeset(
...
         .filter_map(|(i, x)| x.map(|x| (i, x)))
-        .map(|(i, j)| (from[i].take().unwrap(), to[j].take().unwrap()))
-        .collect::<Vec<_>>();
-
-    let new_tasks = to.into_iter().flat_map(|x| x);
+        .map(|(i, j)| {
+            let from = from[i].take().unwrap();
+            let to = to[j].take().unwrap();
+            let delta = if from.tags.get("rec").is_some() {
+                Recurred(vec![to])
```
[the remark this answers is not in the document]
Comment (Nadrieril) [reply src/task_change.rs]: The filtering is indeed done later. I don't want to rename it because once it has been outputted to the user, it is indeed correctly capturing tasks that have been recurred.

File: src/task_change.rs [hunk 7]
```diff
@@ -512,8 +594,11 @@ pub fn display_changeset(
     // Nice display
     if no_changes {
         println!("No changes.");
+        return;
```
[the remark this answers is not in the document]
Comment (Nadrieril) [reply src/task_change.rs]: Any trivial amount of testing after adding such a new functionality would catch that, wouldn't it ?

File: src/lib.rs [hunk 8]
```diff
@@ -11,7 +11,17 @@ extern crate todo_txt;
...
+#[macro_use]
+extern crate serde_derive;
 
 pub mod compute_changes;
 pub mod display_changes;
 pub mod stable_marriage;
+
+#[cfg(all(test, not(feature = "integration_tests")))]
+#[test]
+// #[ignore]
```
[the remark this answers is not in the document]
Comment (Nadrieril) [reply src/lib.rs]: It is there to make it easy to ignore this potentially annoying test without needing to go through the documentation to remember what the syntax is to ignore tests

File: tests/changeset_tests.rs [hunk 9]
```diff
@@ -1,172 +1,83 @@
...
-    let (new_tasks, changes) = compute_changeset(from, to, 0);
-
-    assert_eq!(
-        new_tasks,
-        tasks_from_strings(vec!["x do a thing", "x do a thing"])
-    );
-    assert_eq!(changes, changes_from_strings(vec![("do a thing", Deleted)]));
+fn read_yaml(path: &str) -> BTreeMap<String, Test> {
+    let file = File::open(path).expect(&format!("Unable to open file ‘{}’", path));
+    serde_yaml::from_reader(BufReader::new(&file)).unwrap_or_else(|e| panic!("{}", e))
```
[the remark this answers is not in the document]
Comment (Nadrieril) [reply tests/changeset_tests.rs]: No, using unwrap prints "paniqued at unwrap", whereas this only prints the parse error in a nice way

File: tests/tests.yaml [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+
+  changes:
+    - Changed: # do a thing
+      - Finished(true)
+    - Recurred: # eat vegetables due:2010-01-01 rec:+1d
+      - # x eat vegetables due:2010-01-01 rec:+1d
+        - Finished(true)
+      - # x eat vegetables due:2010-01-02 rec:+1d
+        - RecurredStrict
+        - Finished(true)
```
[the remark this answers is not in the document]
Comment (Nadrieril) [reply tests/tests.yaml]: You're welcome to add an issue. This is a new feature we didn't have before so I don't see why it would be relevant to add in this PR

File: src/stable_marriage.rs (hunk 1 above)
Comment (Nadrieril) [reply src/stable_marriage.rs]: No, both Item and Target can be Man or Woman. The algorithm uses two matchers, one where Item=Man and Target=Woman, and another that has those two swapped. The naming is on purpose different

File: src/stable_marriage.rs (hunk 3 above)
Comment (Nadrieril) [reply src/stable_marriage.rs]: The two are identical, and this algorithm follows the standard Gale-Shapley algorithm you can find anywhere (modulo handling of allowed_divergence and having unmatched individuals).

File: src/stable_marriage.rs (hunk 5 above)
Comment (Nadrieril) [reply src/stable_marriage.rs]: You're right, I didn't know about ranges before three days ago ^^

File: src/stable_marriage.rs (hunk 4 above)
Comment (Nadrieril) [reply src/stable_marriage.rs]: Indeed ^^

File: src/task_change.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Ekleog) [reply src/task_change.rs]: What I mean is that currently this test passes:
```yaml
recurrable_changed:
  allowed_divergence: 40

  from:
    - hello rec:+1d

  to:
    - hella rec:+1d

  new: []

  changes:
    - Recurred:
      -
        - Subject("hello", "hella")
```

Which in my current understanding of the code, I take to mean that `Recurred` actually means “recurrable”?

As for the output to the user, I can't find the place where there's another layer of filtering… maybe it's in one of the follow-up PRs I haven't yet looked at? All I'm finding is this where the two are treated almost the same, and from my understanding from there on they are used only as iterators?
```rust
Changed(t) => Changed(changes_between(&x.orig, &t, true)),
Recurred(changes) => {
    // TODO: compute changes more cleverly
    let changes = changes
        .iter()
        .enumerate()
        .map(|(i, to)| changes_between(&x.orig, to, i == 0))
        .collect::<Vec<_>>();
    Recurred(changes)
}
```

File: src/task_change.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (Ekleog) [reply src/task_change.rs]: Well I'm not really sure, if eg. I add category A, and then test with both new tasks and a category A change, then I won't hit the failing code path (as the added new task would make `no_changes` be `false`).

And I must say I don't really see short-cutting 3 `if`s and a sort of an empty vector an advantage big enough to warrant this risk, however small it is, so long as Rust doesn't come with a formal proof framework strong enough to ensure that such a code path doesn't exist.

That said, upon more thought, maybe there'd be a way to avoid both unneeded computation and this code path:
1. set `no_changes = true`
2. inside each `if`, set `no_changes = false`
3. put this `if no_changes` test after all the `if`s

File: src/lib.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (Ekleog) [reply src/lib.rs]: Oh, didn't think about it. In this case, maybe change the `#[cfg]` above to
```rust
#[cfg(all(test,
          not(feature="integration_tests"),
          not(feature="ignore_integration_tests_warning")))]
```
and adapt the message in the `panic` to mention it? This would make it even easier to ignore this check :)

File: tests/tests.yaml (hunk 10 above)
[the remark this answers is not in the document]
Comment (Ekleog) [reply tests/tests.yaml]: Indeed, I wasn't trying to suggest this should be done right now, hence the idea of adding a `# TODO`, in the same way as you did for the other “this is currently suboptimal behaviour” tests. That said if you don't see the point in such a feature I'll just add it myself later, this way your name won't be on the blame in case someone will want to know who added that TODO :)

File: src/stable_marriage.rs (hunk 1 above)
Comment (Ekleog) [reply src/stable_marriage.rs]: Hmm OK. I guess I'll have to re-read the code to understand how it works then, thanks :)

File: src/task_change.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Nadrieril) [reply src/task_change.rs]: Ah, you're right, it seem that I only fix this in the next PR

File: src/task_change.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (Nadrieril) [reply src/task_change.rs]: I like this, and it turns out this is redundant with the `is_first` variable

File: tests/changeset_tests.rs [hunk 11]
```diff
@@ -1,172 +1,83 @@
...
 use todiff::compute_changes::*;
-use todo_txt::Date as TaskDate;
 use todo_txt::Task;
 
-fn tasks_from_strings(strings: Vec<&str>) -> Vec<Task> {
+fn tasks_from_strings(strings: Vec<String>) -> Vec<Task> {
     strings
         .into_iter()
-        .map(|x| Task::from_str(&x).unwrap())
+        .flat_map(|s| Task::from_str(&s))
```
[the remark this answers is not in the document]
Comment (Nadrieril) [reply tests/changeset_tests.rs]: No you're right

File: tests/changeset_tests.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (Nadrieril) [reply tests/changeset_tests.rs]: `unwrap()` displays 'paniqued at unwrap() [...]' whereas this displays only the serde parse error.

File: src/stable_marriage.rs (hunk 3 above)
Comment (Nadrieril) [reply src/stable_marriage.rs]: The `find_perfect_match` function only looks at women that do not already have a match, so this shouldn't happen. I added a comment

File: src/task_change.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Nadrieril) [reply src/task_change.rs]: I changed that

**Comment by Nadrieril** [comment]
I took all your comments into account :)

File: src/task_change.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Ekleog) [reply src/task_change.rs]: Great, thanks!

File: src/task_change.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (Ekleog) [reply src/task_change.rs]: Oh indeed, I hadn't noticed :+1:

File: tests/tests.yaml (hunk 10 above)
[the remark this answers is not in the document]
Comment (Ekleog) [reply tests/tests.yaml]: Hmm… Currently the display side of things is literally only copying the `Changes` in plain English, with the only added logic of pretty-printing, so that'd go against current program structure (and impede translation should we ever want to do one).

Anyway, added an issue so that it can be tracked :)

**Comment by Ekleog** [comment]
Great, thanks! Merged :)
